text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+Ember addons do respect node_module resolution for build-time code, but they do not respect it for runtime code. This is an unhelpful distinction.
+
+### Verbose, Static Javascript as a Compiler Target
+
+Ember’s strong conventions mean that many kinds of dependencies can be inferred (including _statically_ inferred) without requiring the developer to laboriously manage them. This is a good thing and I believe the current fad in the wider Javascript ecosystem for making developers hand-write verbose static imports for everything confuses the benefits of having static analysis (which is good) with the benefits of hand-managing those static imports (which is unnecessary cognitive load when you have clear conventions and a compiler).
+
+This design is about compiling today’s idiomatic Ember code into more “vanilla” patterns that leverage ES modules, node_modules resolution, and spec-compliant static and dynamic `import` to express the structure of an Ember application in a much more “vanilla Javascript” way.
+
+This compile step lets us separate the authoring format (which isn’t changing in any significant way in this RFC) from the packaging format (which can be more verbose and static than we would want in an authoring format).
+
+# Detailed design
+
+## Definitions
+
+**package**: every addon and app is a package. Usually synonymous with “NPM package”, but we also include in-repo packages. The most important fact about a package is that it’s often the boundary around code that comes from a particular author, team, or organization, so coordination across packages is a more sensitive design problem than coordination within apps.
+
+**app**: a package used at the root of a project.
+
+**addon**: a package not used at the root of a project. Will be an **allowed dependency** of either an **app** or an **addon**.
+
+**allowed dependency**: For **addons**, the **allowed dependencies** are the `dependencies` and `peerDependencies` in `package.json` plus any in-repo addons. For **apps**, the **allowed dependencies** are the `dependencies`, `peerDependencies`, and `devDependencies` in `package.json` plus any in-repo addons.
+
+**Ember package metadata**: the `ember-addon` section inside `package.json`. This already exists in v1, we’re going to extend it.
+
+**v2 package**: a package with `package.json` like:
+
+    "keywords": [ "ember-addon" ],
+    "ember-addon": {
+      "version": 2
+    }
+
+**v1 package**: a package with `package.json` like:
+
+    "keywords": [ "ember-addon" ]
+
+and no `version` key (or version key less than 2) in **Ember package metadata**.
+
+**non-Ember package**: a package without `keywords: ["ember-addon"]`
+
+## Scope of this RFC
+
+This RFC is intended as the base level spec for v2 Ember packages. **It does not attempt to cover everything a v1 package can do today**. For example, no provision is made in this RFC for:
+
+- providing dev middleware
```

Comment:
Out of curiousity. If everyone seriously objected to one or more of these following RFCs, would it disrupt the whole system? And leave the implementation of this RFC in a difficult spot? Or can this RFC exist on its own without one of these pieces?

Reply:
It can exist by itself. In practice it would mean that v1 and v2 packages live side-by-side indefinitely because v2 never finishes implementing all the things v1 packages can do.